### PR TITLE
adds IRGen for more graph_op features

### PR DESCRIPTION
**Pull request by marcrasi**
This PR adds IRGen for:
* `tfc.configure*` device-configuration ops (this involved moving a lot of device configuration code out into a new header and source file so that IRGen can access it)
* function attributes (but does not actually put the functions into the graph at runtime, so this does not work yet)
* shapelist attributes
* typelist attribures
* graph_ops with multiple results

This PR also fixes two bugs exposed by new tests:
* deabstraction was not setting `fn->TFDeabstracted` when it successfully early-exited.
* deabstraction was not fully deabstracting accelerator-only functions in dynamic mode.

This PR enables dynamic compilation on many newly-passing tests. I ifdef'd out the dataset tests that use function attributes because those don't work fully yet.

### CODE REVIEW

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Review by mhong — APPROVED** [review]
nice code structuring!

File: include/swift/Runtime/RuntimeFunctions.def [hunk 1]
```diff
@@ -1327,6 +1327,11 @@ FUNCTION(TFE_OpSetAttrType, TFE_OpSetAttrType, C_CC,
          ARGS(Int8PtrTy, Int8PtrTy, Int32Ty),
          ATTRS(NoUnwind))
 
+FUNCTION(TFE_OpSetAttrFunctionName, TFE_OpSetAttrFunctionName, C_CC,
```
[the remark this answers is not in the document]
Comment (mhong) [reply include/swift/Runtime/RuntimeFunctions.def]: Update: I find it now. it was added recently.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
https://github.com/ipython-contrib/jupyter_contrib_nbextensions

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@bgogul, "sese_loop_canonicalization.swift" is failing on Mac CI for this PR. Looks like it's also failing on Mac CI for master (https://ci-external.swift.org/job/oss-swift-RA-macOS-tensorflow/206/console). So I'll submit this PR despite the failure.

The error is:
```
[ RUN      ] ControlFlow.powerOfTwo_CPU_OR_GPU
[       OK ] ControlFlow.powerOfTwo_CPU_OR_GPU
[ RUN      ] ControlFlow.natSumWithBreak_CPU_OR_GPU
[       OK ] ControlFlow.natSumWithBreak_CPU_OR_GPU
[ RUN      ] ControlFlow.sumOfProducts_CPU_OR_GPU
[       OK ] ControlFlow.sumOfProducts_CPU_OR_GPU
[ RUN      ] ControlFlow.sumOfProductsWithBound_CPU_OR_GPU
[       OK ] ControlFlow.sumOfProductsWithBound_CPU_OR_GPU
[ RUN      ] ControlFlow.doWhileLoopWithBreak_CPU_OR_GPU
the test crashed unexpectedly
[     FAIL ] ControlFlow.doWhileLoopWithBreak_CPU_OR_GPU
[ RUN      ] ControlFlow.nestedDoWhileLoopWithBreak_CPU_OR_GPU
[       OK ] ControlFlow.nestedDoWhileLoopWithBreak_CPU_OR_GPU
ControlFlow: Some tests failed, aborting
UXPASS: []
FAIL: ["doWhileLoopWithBreak_CPU_OR_GPU"]
SKIP: []
To debug, run:
$ /Users/swiftci/jenkins/workspace/swift-PR-TensorFlow-macOS/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/test-macosx-x86_64/TensorFlowRuntime/Output/sese_loop_canonicalization.swift.tmp/a.out --stdlib-unittest-in-process --stdlib-unittest-filter "doWhileLoopWithBreak_CPU_OR_GPU"
```
We might have to run it on one of our own macs to get more detailed information.

**Pull request merged**